opm/simulators/linalg/bda/rocalutionSolverBackend.hpp

```diff
@@ -0,0 +1,108 @@
+/*
+  Copyright 2022 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_ROCALUTIONSOLVER_BACKEND_HEADER_INCLUDED
+#define OPM_ROCALUTIONSOLVER_BACKEND_HEADER_INCLUDED
+
+#include <opm/simulators/linalg/bda/BdaResult.hpp>
+#include <opm/simulators/linalg/bda/BdaSolver.hpp>
+#include <opm/simulators/linalg/bda/WellContributions.hpp>
+
+namespace rocalution {
+template<class Matrix, class Vector, class Scalar> class BiCGStab;
+template<class Matrix, class Vector, class Scalar> class ILU;
+template<class Scalar> class LocalMatrix;
+template<class Scalar> class LocalVector;
+}
+
+namespace Opm
+{
+namespace Accelerator
+{
+
+/// This class implements a rocalution based linear solver solver on GPU
+/// It uses ilu0-bicgstab
+template <unsigned int block_size>
+class rocalutionSolverBackend : public BdaSolver<block_size>
+{
+    typedef BdaSolver<block_size> Base;
+
+    using Base::N;
+    using Base::Nb;
+    using Base::nnz;
+    using Base::nnzb;
+    using Base::verbosity;
+    using Base::platformID;
+    using Base::deviceID;
+    using Base::maxit;
+    using Base::tolerance;
+    using Base::initialized;
+
+private:
+    std::vector<double> h_x; // store solution vector on host
+    int *tmp_rowpointers;    // store matrix on host, this pointer is given to and freed by rocalution
+    int *tmp_colindices;     // store matrix on host, this pointer is given to and freed by rocalution
+    double *tmp_nnzvalues;   // store matrix on host, this pointer is given to and freed by rocalution
+
+    std::unique_ptr<rocalution::BiCGStab<rocalution::LocalMatrix<double>, rocalution::LocalVector<double>, double> > roc_solver;
+    std::unique_ptr<rocalution::ILU<rocalution::LocalMatrix<double>, rocalution::LocalVector<double>, double> > roc_prec;
+
+    /// Initialize sizes and allocate memory
+    /// \param[in] matrix     matrix A
+    void initialize(std::shared_ptr<BlockedMatrix> matrix);
+
+    /// Convert matrix to rocalution format
+    /// copy matrix to raw pointers, which are given to and freed by rocalution
+    /// \param[in] matrix     matrix A
+    void convert_matrix(std::shared_ptr<BlockedMatrix> matrix);
```

Comment:
Is there a special reason to use shared_ptr here? Matrix is not stored inside of these methods and one could also use a referece to the object instead. This is probably nitpicking/personal taste.

Reply:
No reason, I simplified the parameters.

Reply:
Cool, thanks.